### PR TITLE
8340572: ConcurrentModificationException when sorting ArrayList sublists

### PR DESCRIPTION
**Pull request by szegedi**
Fixes a regression with #17818 where `ArrayList.subList(…).sort()` started incrementing `ArrayList.modCount` resulting in some cases throwing a `ConcurrentModificationException` where none was thrown before.

This change keeps the optimization from #17818 but restores the behavior where only sorting the `ArrayList` changes the mod count, but sorting its sublists does not.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340572](https://bugs.openjdk.org/browse/JDK-8340572): ConcurrentModificationException when sorting ArrayList sublists (**Bug** - P2)

### Reviewers
 * [Stuart Marks](https://openjdk.org/census#smarks) (@stuart-marks - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21250/head:pull/21250` \
`$ git checkout pull/21250`

Update a local copy of the PR: \
`$ git checkout pull/21250` \
`$ git pull https://git.openjdk.org/jdk.git pull/21250/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21250`

View PR using the GUI difftool: \
`$ git pr show -t 21250`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21250.diff">https://git.openjdk.org/jdk/pull/21250.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21250#issuecomment-2381442263)

### CODE REVIEW

**Comment by dholmes-ora** [comment]
Just an observation, but sorting is not defined as a "structural modification" but obviously would interfere with an active iterator. So the docs may need updating to include this aspect.

**Comment by dfuch** [comment]
I am not a spcialist here. But if sorting the parent list is considered as a modification, and if the sublist is just a view of the parent list, then surely sorting/modifying the sublist should be considered as a modification of the parent list? 

In which case the issue here is probably deeper: SubList and parent list appear to maintain separate `modcount` variables which are only loosely coupled. Maybe that's the bug that ought to be fixed?

**Comment by dholmes-ora** [comment]
I agree with just reverting the behaviour as discussed in JBS, but will leave the actual review for the libs folk. The more general question of when to trigger CME is a very thorny one.

**Comment by stuart-marks** [comment]
Right, @dholmes-ora, you might recall that we argued about this several years ago. :-) Even though I'm on the losing end of that argument, I am ok with eventually removing the modCount increments from sort and replaceAll. This is covered by JDK-8203704 which should include a spec clarification to support the change in behavior. Meanwhile, this PR should simply revert to the behavior previous to the JDK-8325679 change.

**Comment by stuart-marks** [comment]
@szegedi Thanks for starting this. Just a couple comments on the test.

1. It would be good to have a test case that verifies that sorting the entire list **does** increment modcount, in addition to the existing test case that tests that sorting a sublist **does not** increment modcount. (As before, this is arguably not the correct behavior, but it's the behavior that the system exhibited prior to JDK-8325679.)
2. It would be good for the test to have a more descriptive name than just the bug ID. Something like SortingModCount or similar would be fine.

Thanks.

File: test/jdk/java/util/ArrayList/Bug8340572.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @test
+ * @bug 8340572
+ * @summary ConcurrentModificationException when sorting ArrayList sublists
+ */
+
+public class Bug8340572 {
```
Comment (liach) [inline test/jdk/java/util/ArrayList/Bug8340572.java:34]: Can we give this a descriptive name, like `ConcurrentSubListSort`?

**Comment by szegedi** [comment]
@stuart-marks and @liach thanks, those are good suggestions – done.

**Review by stuart-marks — APPROVED** [review]
Thanks for the updates! I've run these changes through our internal testing system and the results look good.

**Pull request closed without merging**